Feat/multi notary (#90)
sse-secure-systems/connaisseur: setup/README.md

#### Configure Notary

> In `helm/values.yaml`, we first need to specify the location of the to be used notary instance. Connaisseur supports the configuration of multiple notary instances, should your image signatures be at different locations. Per default the official DockerHub notary is configured, with the public key for all official images, but for now let's assume you want to add a new notary instance. Simply add a new entry in the `notary` field, specifying a `name`, `host` and a list of public root keys (`rootKeys`). The `name` should be unique among all notary configurations, the `host` points to the host address of the notary instance and the list of public root keys will be used to verify the image signatures residing in this notary instance. Why a list, you may ask? Just take a look at the the official DockerHub notary, where the official images are signed with a different key as your own images you can push to DockerHub. If you want to verify both, official images and your own (all residing on DockerHub), you'll need multiple keys. How to obtain these will be described further below ([here](#retrieving-the-public-root-key)).

Reviewer: ```suggestion
In `helm/values.yaml`, we first need to specify the location of the notary instance to be used. Connaisseur supports the configuration of multiple notary instances, should your image signatures reside at different locations. Per default, the official DockerHub notary is configured, with the public key for all official images, but for now let's assume you want to add a new notary instance. Simply add a new entry in the `notary` field, specifying a `name`, `host` and a list of public root keys (`rootKeys`). The `name` should be unique among all notary configurations, the `host` points to the host address of the notary instance and the list of public root keys will be used to verify the image signatures residing in this notary instance. Why a list, you may ask? Just take a look at the the official DockerHub notary, where the official images are signed with a different key as your own images you can push to DockerHub. If you want to verify both, official images and your own (all residing on DockerHub), you'll need multiple keys. How to obtain these will be described further below ([here](#retrieving-the-public-root-key)).
```

> In addition to these three mandatory settings, a self-signed certificate and authentication can be added, as well as marking a notary belonging to a azure container registry (acr). The self-signed certificate can be added with the `selfsignedCert` field, where it will be pasted in a PEM format. The authentication detail can be defined in the `auth` field, where either the credentials are given directly with `auth.user` and `auth.password` or a predefined [Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/) is referenced in the `auth.secretName` field. The data of the predefined secret needs to be a `cred.yaml` with the fields `USER` and `PASS`. Lastly the `isAcr` can be used to mark the notary as being part of an azure container registry, as these types of notaries present a slightly different kind of behavior. All in all a sample notary configuration might look like this:

Reviewer: ```suggestion
In addition to these three mandatory settings, a self-signed certificate and authentication can be added, as well as marking a notary belonging to an Azure Container Registry (ACR). The self-signed certificate can be added with the `selfsignedCert` field, where it will be pasted in PEM format. The authentication detail can be defined in the `auth` field, where either the credentials are given directly with `auth.user` and `auth.password` or a predefined [Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/) is referenced in the `auth.secretName` field. The data of the predefined secret needs to be a `cred.yaml` with the fields `USER` and `PASS`. Lastly the `isAcr` can be used to mark the notary as being part of an Azure Container Registry, as these types of notaries present a slightly different kind of behavior. All in all a sample notary configuration might look like this:
```

```yaml
notary:
- name: harbor
  host: notary.harbor.domain
  rootKeys:
  - name: library
    key: |
      -----BEGIN PUBLIC KEY-----
      ...
      -----END PUBLIC KEY-----
  - name: app1
    key: | 
      -----BEGIN PUBLIC KEY-----
      ...
      -----END PUBLIC KEY-----
  selfsignedCert: | 
    -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE-----
  auth:
    user: notaryuser
    password: notarypassword
    # secretName: notary-secret <-- alternatively
    # the secret data needs to look like this:
    # data:
    #   cred.yaml: b46enc(USER: notaryuser\nPASS: notarypassword)
  # isAcr: true 
```

If you simply want to verify non-official images on DockerHub, then just add the corresponding public key to the existing predefined notary configuration under `rootKeys`, next to the key for official images. But how do you get these public root keys?

#### Retrieving the public root key

> For now let's assume you only want to verify image signatures you created/will create, using the key generated in step [two](#set-up-docker-content-trust). To get the public part of this you have to convert the key manually.

Reviewer: ```suggestion
For now let's assume you only want to verify image signatures you created/will create, using the key generated in [step two](#set-up-docker-content-trust). To get the public part of this you have to convert the key manually.
```

Author: resolved in 3b7d64cd6e6c5c09c2f7370d57ddb8a3b6c8869c

- Go to your private docker trust folder `~/.docker/trust/private` and identify the root key. To do so, you can either go through the files and search for the one with `role: root` or `grep -r 'role: root'`.
- Copy the root key file to a new file `root-priv.key` and remove line with `role: root` and the following empty line.
- Generate the corresponding public key via `openssl ec -in root-priv.key -pubout -out root.pub`. You will be asked to enter your root password.
- The new `root.pub` contains your public key which you copy and set as a new `rootKeys` entry in your notary configuration of choice in `connaisseur/helm/values.yaml`. The result should look similar as in the example above.
- To clean up, remove the `root-priv.key` and `root.pub` in `~/.docker/trust/private`.

Getting the public key can also be summarized in a small script:

```bash
cd ~/.docker/trust/private
sed '/^role:\sroot$/d' $(grep -iRl "role: root$" .) > root-priv.key
openssl ec -in root-priv.key -pubout -out root.pub
```

#### Configure the image policy

> Since we have potentially multiple notary instances and multiple keys for each instance, Connaisseur somehow needs to know for which images it should use what kind of combination of instance and key. This is configured in the `policy` part of the `helm/values.yaml`, where different rules are defined, that need to be matched against a given image. Only one rule will be matched at the same time and this rule can specify which notary should be used, with the `notary` field. The value of this field has to match one of the notary `name` fields. Additionally a `key` field can be given in the rule, which will then specify which key within the chosen notary configuration should be used. This again has to match a value of a `name` field of one of the `rootKeys` entries. If no `notary` or `key` fields are given inside a rule, then the top most entry will be chosen respectively.

Reviewer: ```suggestion
Since we have potentially multiple notary instances and multiple keys for each instance, Connaisseur somehow needs to know for which images it should use what kind of combination of instance and key (otherwise, any key compromised in your supply chain would constitute a (potential) compromise of all your images). This is configured in the `policy` part of the `helm/values.yaml`, where different rules are defined, that need to be matched against a given image. Only one rule will be matched at the same time and this rule can specify which notary should be used, with the `notary` field. The value of this field has to match one of the notary `name` fields. Additionally a `key` field can be given in the rule, which will then specify which key within the chosen notary configuration should be used. This again has to match a value of a `name` field of one of the `rootKeys` entries. If no `notary` or `key` fields are given inside a rule, then the top most entry will be chosen respectively.
```

Author: resolved in 3b7d64cd6e6c5c09c2f7370d57ddb8a3b6c8869c

> Since we have potentially multiple notary instances and multiple keys for each instance, Connaisseur somehow needs to know for which images it should use what kind of combination of instance and key. This is configured in the `policy` part of the `helm/values.yaml`, where different rules are defined, that need to be matched against a given image. Only one rule will be matched at the same time and this rule can specify which notary should be used, with the `notary` field. The value of this field has to match one of the notary `name` fields. Additionally a `key` field can be given in the rule, which will then specify which key within the chosen notary configuration should be used. This again has to match a value of a `name` field of one of the `rootKeys` entries. If no `notary` or `key` fields are given inside a rule, then the top most entry will be chosen respectively.

Reviewer: If we choose differently in the ADR, the last sentence will need to be changed

> In the below example, all images coming from a `core.harbor.domain` registry will be verified using the `harbor` notary from the example above and use it's `app1` key. All other images take the predefined `dockerhub` notary and it's `library` key.

Reviewer: ```suggestion
In the below example, all images coming from a `core.harbor.domain` registry will be verified using the `harbor` notary from the example above and use its `app1` key. All other images take the predefined `dockerhub` notary and its `library` key.
```

Author: resolved in 3b7d64cd6e6c5c09c2f7370d57ddb8a3b6c8869c

## Cleanup

> If you want to remove Connaisseur, run `make uninstall`.

Reviewer: Why are you dropping the namespace cleanup hint again?

Author: cuz of this:
https://github.com/sse-secure-systems/connaisseur/pull/90#discussion_r576136886

## Cleanup

> If you want to remove Connaisseur, run `make uninstall`. We do not delete the namespace, since you might have deployed things in there, which we would inadvertently delete. If you want to get rid of the namespace, run `kubectl delete ns WHATEVEROURDEFAULTIS`.

Reviewer: ```suggestion
If you want to remove Connaisseur, run `make uninstall`. We do not delete the namespace, since you might have deployed things in there, which we would inadvertently delete. If you want to get rid of the namespace, run `kubectl delete ns connaisseur` (or whatever other namespace you installed Connaisseur to).
```
